### PR TITLE
python3Packages.testbook: init at 0.4.2

### PR DESCRIPTION
**Pull request by djacu**
###### Description of changes

Add python package testbook.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by jlesquembre** [comment]
Result of `nixpkgs-review pr 197551` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>python310Packages.testbook</li>
    <li>python39Packages.testbook</li>
  </ul>
</details>

**Comment by djacu** [comment]
I tested this in a Python shell with versions 3.9 and 3.10. Not sure which "tested" box to check off in the PR template.

File: pkgs/development/python-modules/testbook/default.nix [hunk 1]
```diff
@@ -0,0 +1,47 @@
+{
+  lib,
```
Comment (fabaff) [inline pkgs/development/python-modules/testbook/default.nix:2]:
```suggestion
{ lib,
```

File: pkgs/development/python-modules/testbook/default.nix [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  pythonOlder,
+  fetchFromGitHub,
+  nbformat,
+  nbclient,
+  ipykernel,
+  pandas,
+  pytest,
+  traitlets,
+}:
+buildPythonPackage rec {
```
Comment (fabaff) [inline pkgs/development/python-modules/testbook/default.nix:13]:
```suggestion
}:

buildPythonPackage rec {
```

File: pkgs/development/python-modules/testbook/default.nix [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+  propagatedBuildInputs = [
+    nbclient
+    nbformat
+  ];
+
+  checkInputs = [
+    ipykernel
+    pandas
+    pytest
```
Comment (fabaff) [inline pkgs/development/python-modules/testbook/default.nix:35]: At first glance it looks like that `pytestCheckHook` could be used. Could you please check that?

```suggestion
    pytestCheckHook
```

File: pkgs/development/python-modules/testbook/default.nix [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    traitlets
+  ];
+
+  checkPhase = "pytest";
```
Comment (fabaff) [inline pkgs/development/python-modules/testbook/default.nix:39]: Bit needed if `pytestCheckHook` works.

```suggestion
```

File: pkgs/development/python-modules/testbook/default.nix [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
```
Comment (fabaff) [inline pkgs/development/python-modules/testbook/default.nix:40]: `pythonImportsCheck` is an additional safe-guard.

```suggestion
  pythonImportsCheck = [
    "testbook"
  ];

```

File: pkgs/development/python-modules/testbook/default.nix [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+  meta = with lib; {
+    description = "A unit testing framework extension for testing code in Jupyter Notebooks";
+    homepage = "https://testbook.readthedocs.io/";
+    license = with licenses; [bsd3];
```
Comment (fabaff) [inline pkgs/development/python-modules/testbook/default.nix:44]: `nixpkgs-fmt` helps to catch stuff like that.

```suggestion
    license = with licenses; [ bsd3 ];
```

File: pkgs/development/python-modules/testbook/default.nix [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    maintainers = with maintainers; [djacu];
```
Comment (fabaff) [inline pkgs/development/python-modules/testbook/default.nix:45]:
```suggestion
    maintainers = with maintainers; [ djacu ];
```

File: pkgs/development/python-modules/testbook/default.nix (hunk 1 above)
Comment (djacu) [reply pkgs/development/python-modules/testbook/default.nix]: Resolved in latest commit

File: pkgs/development/python-modules/testbook/default.nix (hunk 2 above)
Comment (djacu) [reply pkgs/development/python-modules/testbook/default.nix]: Resolved in latest commit

File: pkgs/development/python-modules/testbook/default.nix (hunk 3 above)
Comment (djacu) [reply pkgs/development/python-modules/testbook/default.nix]: Yup it works fine.

File: pkgs/development/python-modules/testbook/default.nix (hunk 4 above)
Comment (djacu) [reply pkgs/development/python-modules/testbook/default.nix]: Resolved in latest commit

File: pkgs/development/python-modules/testbook/default.nix (hunk 5 above)
Comment (djacu) [reply pkgs/development/python-modules/testbook/default.nix:42]: Works fine along with  `pytestCheckHook`

File: pkgs/development/python-modules/testbook/default.nix (hunk 6 above)
Comment (djacu) [reply pkgs/development/python-modules/testbook/default.nix]: Resolved in latest commit

File: pkgs/development/python-modules/testbook/default.nix (hunk 7 above)
Comment (djacu) [reply pkgs/development/python-modules/testbook/default.nix]: Resolved in latest commit

**Comment by djacu** [comment]
I wasn't able to do a batch commit from the review so I formatted with `nixpkgs-fmt` and pushed a new commit. Also switched to using `pytestCheckHook` and `pythonImportsCheck`. Both work fine in a `nix-shell`.

**Review by garbas — APPROVED** [review]

**Pull request merged**